File: java-automation/src/main/java/com/automation/pages/InventoryPage.java
package com.automation.pages;

import com.microsoft.playwright.Page;
import com.microsoft.playwright.Locator;
import com.microsoft.playwright.options.AriaRole;
import com.automation.data.TestData;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

public class InventoryPage {
    private final Page page;

    public InventoryPage(Page page) {
        this.page = page;
    }

    private Locator getProductItems() {
        return page.locator(".inventory_item");
    }

    private Locator getProductNames() {
        return page.locator(".inventory_item_name");
    }

    private Locator getProductPrices() {
        return page.locator(".inventory_item_price");
    }

    private Locator getCartBadge() {
        return page.locator(".shopping_cart_badge");
    }

    private Locator getCartLink() {
        return page.locator(".shopping_cart_link");
    }

    private Locator getSortDropdown() {
        return page.locator("[data-test='product-sort-container']");
    }

    private Locator getBurgerMenuButton() {
        return page.getByRole(AriaRole.BUTTON, new Page.GetByRoleOptions().setName("Open Menu"));
    }

    private Locator getResetAppStateLink() {
        return page.getByRole(AriaRole.LINK, new Page.GetByRoleOptions().setName("Reset App State"));
    }

    private Locator getLogoutLink() {
        return page.getByRole(AriaRole.LINK, new Page.GetByRoleOptions().setName("Logout"));
    }

    public void goTo() {
        page.navigate(TestData.Urls.INVENTORY);
    }

    public void sortBy(String option) {
        getSortDropdown().selectOption(option);
    }

    /** Backwards-compatible alias kept for existing callers */
    public void selectSort(String option) {
        sortBy(option);
    }

    public void addToCartByDataTest(String dataTestId) {
        page.locator("[data-test='" + dataTestId + "']").click();
    }

    public void addMultipleToCart(List<String> dataTestIds) {
        for (String id : dataTestIds) {
            addToCartByDataTest(id);
        }
    }

    public void removeFromCartByDataTest(String dataTestId) {
        page.locator("[data-test='" + dataTestId + "']").click();
    }

    public void openProductDetail(String productName) {
        page.getByText(productName, new Page.GetByTextOptions().setExact(true)).click();
    }

    public void goToCart() {
        getCartLink().click();
        assertEquals(TestData.Urls.CART, page.url(), "User should be redirected to cart page");
    }

    public void logout() {
        getBurgerMenuButton().click();
        getLogoutLink().click();
        assertEquals(TestData.Urls.BASE, page.url(), "User should be redirected to login page after logout");
    }

    public void resetAppState() {
        getBurgerMenuButton().click();
        getResetAppStateLink().click();
    }

    // ─── Assertions ──────────────────────────────────────────────────────────────

    public void assertProductCount(int expected) {
        assertEquals(expected, getProductItems().count(),
            "Expected " + expected + " products but found " + getProductItems().count());
    }

    public void assertCartBadgeCount(int expected) {
        assertTrue(getCartBadge().isVisible(), "Cart badge should be visible");
        assertEquals(String.valueOf(expected), getCartBadge().textContent(),
            "Cart badge should show " + expected + " items");
    }

    public void assertCartBadgeNotVisible() {
        assertFalse(getCartBadge().isVisible(), "Cart badge should not be visible");
    }

    private List<Double> parsePrices() {
        List<Double> prices = new ArrayList<>();
        for (String raw : getProductPrices().allTextContents()) {
            prices.add(Double.parseDouble(raw.replace("$", "").trim()));
        }
        return prices;
    }

    public List<Double> assertPricesSortedDescending() {
        List<Double> prices = parsePrices();
        for (int i = 1; i < prices.size(); i++) {
            assertTrue(prices.get(i - 1) >= prices.get(i),
                "Prices not sorted descending: " + prices);
        }
        return prices;
    }

    public List<Double> assertPricesSortedAscending() {
        List<Double> prices = parsePrices();
        for (int i = 1; i < prices.size(); i++) {
            assertTrue(prices.get(i - 1) <= prices.get(i),
                "Prices not sorted ascending: " + prices);
        }
        return prices;
    }

    public List<String> assertNamesAlphabeticallyAscending() {
        List<String> names = new ArrayList<>(getProductNames().allTextContents());
        List<String> sorted = new ArrayList<>(names);
        Collections.sort(sorted);
        assertEquals(sorted, names, "Names should be sorted A to Z");
        return names;
    }

    public List<String> assertNamesAlphabeticallyDescending() {
        List<String> names = new ArrayList<>(getProductNames().allTextContents());
        List<String> sorted = new ArrayList<>(names);
        sorted.sort(Collections.reverseOrder());
        assertEquals(sorted, names, "Names should be sorted Z to A");
        return names;
    }
}

File: python-automation/src/automation/pages/inventory_page.py
from typing import List
from playwright.sync_api import Page
from src.automation.data.test_data import Urls


class InventoryPage:
    def __init__(self, page: Page):
        self.page = page

    @property
    def product_items(self):
        return self.page.locator(".inventory_item")

    @property
    def product_names(self):
        return self.page.locator(".inventory_item_name")

    @property
    def product_prices(self):
        return self.page.locator(".inventory_item_price")

    @property
    def cart_badge(self):
        return self.page.locator(".shopping_cart_badge")

    @property
    def cart_link(self):
        return self.page.locator(".shopping_cart_link")

    @property
    def sort_dropdown(self):
        return self.page.locator("[data-test='product-sort-container']")

    @property
    def burger_menu_button(self):
        return self.page.get_by_role("button", name="Open Menu")

    @property
    def reset_app_state_link(self):
        return self.page.get_by_role("link", name="Reset App State")

    @property
    def logout_link(self):
        return self.page.get_by_role("link", name="Logout")

    def go_to(self):
        self.page.goto(Urls.INVENTORY)

    def sort_by(self, option: str):
        self.sort_dropdown.select_option(option)

    # Backwards-compatible alias for the original method name
    def select_sort(self, option: str):
        self.sort_by(option)

    def add_to_cart_by_data_test(self, data_test_id: str):
        self.page.locator(f"[data-test='{data_test_id}']").click()

    def add_multiple_to_cart(self, data_test_ids: List[str]):
        for data_test_id in data_test_ids:
            self.add_to_cart_by_data_test(data_test_id)

    def remove_from_cart_by_data_test(self, data_test_id: str):
        self.page.locator(f"[data-test='{data_test_id}']").click()

    def open_product_detail(self, product_name: str):
        self.page.get_by_text(product_name, exact=True).click()

    def go_to_cart(self):
        self.cart_link.click()
        assert self.page.url == Urls.CART, \
            f"User should be redirected to {Urls.CART}, but got {self.page.url}"

    def logout(self):
        self.burger_menu_button.click()
        self.logout_link.click()
        assert self.page.url == Urls.BASE, \
            f"User should be redirected to {Urls.BASE}, but got {self.page.url}"

    def reset_app_state(self):
        self.burger_menu_button.click()
        self.reset_app_state_link.click()

    # ─── Accessors ───────────────────────────────────────────────────────────

    def get_product_names(self) -> List[str]:
        return self.product_names.all_text_contents()

    def get_product_prices(self) -> List[float]:
        raw = self.product_prices.all_text_contents()
        return [float(price.replace("$", "")) for price in raw]

    # ─── Assertions ──────────────────────────────────────────────────────────

    def assert_product_count(self, expected: int):
        count = self.product_items.count()
        assert count == expected, \
            f"Expected {expected} products but found {count}"

    def assert_cart_badge_count(self, expected: int):
        assert self.cart_badge.is_visible(), "Cart badge should be visible"
        badge_text = self.cart_badge.text_content()
        assert badge_text == str(expected), \
            f"Cart badge should show {expected} items but shows {badge_text}"

    def assert_cart_badge_not_visible(self):
        assert not self.cart_badge.is_visible(), \
            "Cart badge should not be visible"

    def assert_prices_sorted_descending(self) -> List[float]:
        prices = self.get_product_prices()
        is_sorted = all(prices[i - 1] >= prices[i] for i in range(1, len(prices)))
        assert is_sorted, f"Prices not sorted descending: {prices}"
        return prices

    def assert_prices_sorted_ascending(self) -> List[float]:
        prices = self.get_product_prices()
        is_sorted = all(prices[i - 1] <= prices[i] for i in range(1, len(prices)))
        assert is_sorted, f"Prices not sorted ascending: {prices}"
        return prices

    def assert_names_alphabetically_ascending(self) -> List[str]:
        names = self.get_product_names()
        assert names == sorted(names), \
            f"Names not sorted ascending: {names}"
        return names

    def assert_names_alphabetically_descending(self) -> List[str]:
        names = self.get_product_names()
        assert names == sorted(names, reverse=True), \
            f"Names not sorted descending: {names}"
        return names
